small cpp optimization

breznak opened the pull request:
TP.compute() replace for-loop with memcpy()
Fixes #280

scottpurdy commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -188,16 +188,14 @@ bool Cells4::computeUpdate(UInt cellIdx, UInt segIdx, CStateIndexed& activeState
       newSynapses.reserve(highWaterSize);
     }
-    for (UInt i = 0; i != segment.size(); ++i)
+    for (UInt i=0; i<segment.size(); ++i) //FIXME why we skip segment[0] here?
```
The first iteration of the loop will be `i=0`. I prefer `++i` over `i++`. They have the same effect if they are used alone. When they are used as part of a more complex statement then `++i` will increment and evaluate to the incremented value while `i++` will increment the variable but evaluate to the pre-incremented value.

So I like switching to `i < segment.size()` (I'd prefer spaces around the comparison operator though), don't have a strong opinion on the `++i` vs `i++` but regardless the comment needs to go.

breznak replied:
thanks, I learned something new today, thought the diff between ++i/i++ would affect the for-loop too. All makes sense now.

scottpurdy commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -1036,11 +1029,12 @@ void Cells4::updateInferenceState(const std::vector<UInt> & activeColumns)
   _infActiveStateT1 = _infActiveStateT;
   _infPredictedStateT1 = _infPredictedStateT;
   memcpy(_cellConfidenceT1, _cellConfidenceT, _nCells * sizeof(_cellConfidenceT[0]));
+  memset(_cellConfidenceT, 0, _nCells * sizeof(_cellConfidenceT[0]));
```
Is this (and the memset later) necessary or redundant? If all of the values are set then no need to zero them first. I haven't looked through the code thoroughly yet but it seems like it would be a pretty obvious bug if the values should have been zeroed but weren't.

scottpurdy replied:
Oh I see the comment on the line later that zeros the values. Now there are three different spots that memset `_cellConfidenceT` which seems excessive. Ideally there would just be one spot as before.

breznak replied:
I tried this in f12abc0 which resulted in memleaks and crashes, so I reverted it. Now retrying with 
Anyway, the time is: for-loops >> memset > only-1 memset, at the cost we are not reseting to `0` where the comments say so (to be honest, comments said, but code did otherwise in the orig version :P )

scottpurdy commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -1291,7 +1279,9 @@ void Cells4::compute(Real* input, Real* output, bool doInference, bool doLearnin
     _avgInputDensity = (Real) activeColumns.size();
   } else {
-    _avgInputDensity = 0.99*_avgInputDensity + 0.01 * (Real) activeColumns.size();
+    //TODO remove magic constants. should this be swarmed-over?
```
nit: put a single space between `//` and the comment

scottpurdy commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -1336,37 +1326,7 @@ void Cells4::compute(Real* input, Real* output, bool doInference, bool doLearnin
   // invalidating our indexes.
   memset(output, 0, _nCells * sizeof(output[0])); // most output is zero
 #if SOME_STATES_NOT_INDEXED
-#ifdef NTA_PLATFORM_darwin86
```
I'm not familiar with the history of these optimizations. What impact does this have? Is it a speed-only change or does it impact the actual results? And how much of a speed impact does it have?

breznak replied:
@scottpurdy it's speed-only change, I don't have hw to measure, but i thought it was a removal of a dead code only - I dont think we are now able to build nupic on 32b OSX, which was with CoreDuo CPUs ~2006.

scottpurdy replied:
That seems reasonable to me.

scottpurdy commented:
Done with this round of review. I put some comments in. I am most worried about the impact of removing some of the preprocessor-guarded optimizations.

breznak commented:
@scottpurdy thanks for the review. ready for round 2 ;)

rhyolight commented:
@scottpurdy Back atcha

scottpurdy commented:
@breznak - My only remaining concern is that memset to zero out new array gets called three times. Please see previous inline comment.

@subutai - Do you have any context on the optimizations that are being removed? Seems fine to me to remove untested OS X 32 bit optimizations but curious if you have more background.

subutai commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -2827,42 +2787,15 @@ void Cells4::computeForwardPropagation(CState& state)
-#else
-  const UInt multipleOf4 = 4 * (_nCells/4);
+  //TODO above was multipleOf8 for NON-darwin, here is multipleOf4 !
+  const UInt multipleOf4 = 4 * (_nCells/4); //TODO remove for small penalty, see "tail"
```
Should we use the above multipleOf8 stepping since this condition is for 64 bit systems?

breznak replied:
that was my concern too. 26affc9 should fix it. It worked before because 64bit could work with the 32bit, and noone is using the darwin86 arch

subutai commented:
@scottpurdy @breznak I don't remember, but it looks like these optimizations are for generic 32 bit RISC, and not specific to OS X. So they really should have generic 32 bit processor guards, not darwin86 specific guards. Do we care about optimizing for 32 bit systems such as Raspberry PI?

scottpurdy commented:
> Do we care about optimizing for 32 bit systems such as Raspberry PI?

Well it's more whether we care __enough__ to have the optimizations. I wouldn't go out of my way to remove it personally but if it is cluttering the codebase or making it more difficult to read then I am alright with removing it. I don't have a strong opinion either way.

breznak commented:
please see also: 
https://github.com/breznak/nupic.core/commit/8e1307293229b3bbb6e328fc5d7b580015458f75#diff-4c232a262b217bd7ab4e516856187881R2790

breznak commented:
@scottpurdy @subutai about darwin86 ...noone is actively using it, just fixed a bug which would break builds on it. As Subutai suggested: remove specific, extend generic to 32bit -> #306

breznak commented:
there are a lot of places in `ArrayAlgo.hpp` like https://github.com/numenta/nupic.core/pull/278/files#diff-5fc4121d5c281de47186c9f30e425bdbR78 
where `ifdef linux64 || darwin64` is used. Should I use more generic `ifdef NTA_ARCH64` ? with potential to make the optimizations used on more platforms (aarch64, win64,..), but potentially breaking the unmerged https://github.com/numenta/nupic.core/pull/275 ?

david-ragazzi commented:
> Should I use more generic ifdef NTA_ARCH64 ? with potential to make the optimizations used on more platforms (aarch64, win64,..), but potentially breaking the unmerged #275 ? 

Don't worry about break this, I can fix it later. Actually I'm favor of we separate archicteture from operating systems and discard platform variables!

I suggest we have:
- Architecture variables: `-DNTA_ARCH_32`, `-DNTA_ARCH_64`, `-DNTA_ARCH_ARM`
- OS variables: `-DNTA_OS_LINUX`, `-DNTA_OS_DARWIN`, `-DNTA_OS_WINDOWS`

david-ragazzi commented on src/CMakeLists.txt:
```diff
@@ -58,6 +58,7 @@ endif()
 
 # Compiler `-D*` definitions
 add_definitions(-DNTA_PLATFORM_${PLATFORM}${BITNESS}
+                -DNTA_ARCH_${BITNESS}
```
How about add also NTA_ARCH_DARWIN, NTA_ARCH_LINUX and NTA_ARCH_WINDOWS here:

https://github.com/numenta/nupic.core/blob/master/src/CMakeLists.txt#L40

```
if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
  set(PLATFORM "darwin")
  set(STDLIB "-stdlib=libc++")
  add_definitions(-DNTA_OS_DARWIN)
elseif(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
  set(PLATFORM "windows")
  set(STDLIB "")
  add_definitions(-DNTA_OS_WINDOWS)
elseif(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
  set(PLATFORM "linux")
  set(STDLIB "")
  add_definitions(-DNTA_OS_LINUX)
endif()
```

david-ragazzi commented on src/nupic/math/ArrayAlgo.hpp:
```diff
@@ -5078,7 +4979,7 @@ namespace nupic {
       // skip the asm. 
       if (n1 > 0) { 
 
-  #if defined (NTA_PLATFORM_darwin86) || defined(NTA_PLATFORM_linux32)
```
Will this works for WIndows too?

breznak replied:
I don't know..if it won't, we should add (not win32)

david-ragazzi replied:
> I don't know..if it won't, we should add (not win32)

Yeah, please add this condition to similar pieces of code. Otherwise, this will cause future problems to Windows build.. :-/

david-ragazzi commented on src/nupic/math/ArrayAlgo.hpp:
```diff
@@ -75,7 +75,7 @@ namespace nupic {
 
   // TODO: add asm code for gcc/clang/... on Windows
 
-  #elif defined(NTA_PLATFORM_darwin86) || defined(NTA_PLATFORM_linux32)
```
@breznak This will break the future builds!

```
#if defined(NTA_PLATFORM_win32) && defined(NTA_COMPILER_MSVC)
...
#elif defined(NTA_ARCH_32)
...
#endif
```

If by any of the conditions in `#if defined(NTA_PLATFORM_win32) && defined(NTA_COMPILER_MSVC)` is false, the compiler will can process non-Windows ASM in `#elif defined(NTA_ARCH_32)`!

Maybe you should use:
`#elif !defined(NTA_PLATFORM_win32) && defined(NTA_ARCH_32)`
defined(NTA_PLATFORM_win32)

david-ragazzi commented on src/nupic/math/ArrayAlgo.hpp:
```diff
@@ -5196,7 +5097,7 @@ namespace nupic {
       if (n1 > 0) {
 
-  #if defined(NTA_PLATFORM_darwin86) || defined(NTA_PLATFORM_linux32)
+  #if defined(NTA_ARCH_32)
```
> I don't know..if it won't, we should add (not win32)

Same here.

breznak commented:
Hope I got this merge right.. 

@david-ragazzi I tried to address the windows concerns as you suggested. 
@scottpurdy can you please re-review after the merge? Seems ready

scottpurdy commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -188,16 +188,15 @@ bool Cells4::computeUpdate(UInt cellIdx, UInt segIdx, CStateIndexed& activeState
       newSynapses.reserve(highWaterSize);
     }
-    for (UInt i = 0; i != segment.size(); ++i)
+    for (UInt i = 0; i < segment.size(); ++i) {
```
current convention is open brace on following line

breznak replied:
just for `for` ? shouldn't it be unified? as i see eg in  https://github.com/numenta/nupic.core/pull/278/files#diff-4c232a262b217bd7ab4e516856187881R558 
..anyway, i will change it and we can discuss code-style another time.

scottpurdy commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -2827,13 +2816,14 @@ void Cells4::computeForwardPropagation(CState& state)
   // Zero out previous values
   // Using memset is quite a bit faster on laptops, but has almost no effect
-  // on Neo15!
+  // on Neo15! // TODO shouldn't we optimize for generic (laptop) CPUs rather than
```
I don't think this todo comment in the code is very useful. Please remove it and if you want to open a GH issue then do so.

breznak replied:
Done in https://github.com/numenta/nupic.core/issues/344

scottpurdy commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -2827,13 +2816,14 @@ void Cells4::computeForwardPropagation(CState& state)
   // links from each source cell.  _cellActivity will be set to the total
   // activity coming into a cell.
-#if defined(NTA_ARCH_32) && defined(NTA_OS_DARWIN)
+#ifdef NTA_ARCH_64
```
You changed the behavior here. Why do we want to switch this from 32 bit OS X to all 64 bit platforms?

Also, the comment for the endif doesn't match.

breznak replied:
I believe the 32/64 guard was wrong there (as the code looks same as above in https://github.com/breznak/nupic.core/commit/26affc98c0a8c3c61d47c0d14de039119fc44d47#diff-4c232a262b217bd7ab4e516856187881R1328) , the code would still work, but not at optimal speed. This fixes that and sets correct UInt64 for 64bit OS,...

scottpurdy commented on src/nupic/math/ArrayAlgo.hpp:
```diff
@@ -3786,106 +3786,7 @@ namespace nupic {
       int n2 = (int)(end - start - n1);
 
 
-#if defined(NTA_PLATFORM_darwin86_disabled) || defined(NTA_PLATFORM_linux32_disabled)
```
It isn't clear to me what the intent of this block was for. Can you comment on why you removed it?

breznak replied:
It was a dead code, we did not set the `NTA_PLATFORM_*_disabled` flags anywhere. 
Plus, very similar ASM optimization is applied to 64b right below 
https://github.com/breznak/nupic.core/commit/9810d06dfff8c10626d83d13acd9e5caa4cc6df4?diff=unified#diff-5fc4121d5c281de47186c9f30e425bdbR3798

scottpurdy commented on src/nupic/math/ArrayAlgo.hpp:
```diff
@@ -5241,7 +5142,7 @@ namespace nupic {
                      );
 
-  #elif defined(NTA_ARCH_64)
+  #else //64bit
```
Isn't it better to leave the block explicit on 64 bit? If we have some random other platform it would be better for it to fail then to used some block not intended for the architecture.

scottpurdy replied:
Actually leaving this up to you, see my comment on the PR: https://github.com/numenta/nupic.core/pull/278#issuecomment-74909035

breznak replied:
Right, it's a matter of taste. I read somewhere (and prefer) to set the least limiting guards (most specific). What you say about the new arch is true, on the other hand, eg (I think @oxtopus ) made port to Solaris which defines a specific platform (NTA_OS_Solaris64) ..and it will take advantage from this too (as Solaris is UNIX).

scottpurdy commented on src/nupic/math/ArrayAlgo.hpp:
```diff
@@ -5075,7 +4976,7 @@ namespace nupic {
     // See comments in count_gt. We need both conditional compilation and 
     // SSE_LEVEL check.
-#if defined(NTA_OS_LINUX) || defined(NTA_OS_DARWIN)
+#if not defined(NTA_OS_WINDOWS)
```
There are a handful of cases where you change from `defined(NTA_OS_LINUX) || defined(NTA_OS_DARWIN)` to `not defined(NTA_OS_WINDOWS)`. I think it is better to condition the code blocks on specific platforms rather than anything-but-windows. That way if there is some other platform we don't accidentally have it use code not intended for it.

scottpurdy replied:
Actually leaving this up to you, see my comment on the PR: https://github.com/numenta/nupic.core/pull/278#issuecomment-74909035

scottpurdy commented:
@breznak @david-ragazzi - I just made some recommendations that we keep:

```
defined(NTA_OS_LINUX) || defined(NTA_OS_DARWIN)
```

instead of changing it to:

```
not defined(NTA_OS_WINDOWS)`
```

But I see that David had recommended the opposite. I think it is better to not assume we only have Windows, Linux, and OS X. So I think it just depends on the code. Think about what should happen if we have a different platform. The code changes I recommended reverting might be OK if it gives the right default behavior for new platforms. @breznak - Let me know when you are happy with the preprocessor checks and I will trust that you have chosen reasonable defaults in those conditionals.

breznak commented:
@scottpurdy Thank you for the review. I addressed your comments. For the platform #ifdef guard, I see both points, would still incline to the !Windows variant, as a new platform can likely be some *nix favour, or ARM or something which would require more complete rework. (i dont have a strong opinion though on this)

scottpurdy commented:
@breznak - The !Windows variant is fine, but any time you change it you will need to look at the code and make a case-by-case decision on what the right default behavior is.

Thanks for all of the updates.

:+1: merge when ready
